### ISSUE TITLE
Support of heatmaps displaying

### ISSUE THREAD

**Issue by NShaforostov** [epam/NGB#512]
- [Background](#background)
- [Approach](#approach)
  - [Heatmap file registration](#heatmap-file-registration)
    - [Dendrogram file registration](#dendrogram-file-registration)
    - [Row/column annotation file registration](#rowcolumn-annotation-file-registration)
    - [Cell annotation file registration](#cell-annotation-file-registration)
  - [Heatmap displaying](#heatmap-displaying)
    - [Color scheme configuration](#color-scheme-configuration)
    - [Displaying additional details](#displaying-additional-details)
  - [Additional options](#additional-options)
    - [Displaying at the Summary page](#displaying-at-summary)
    - [Opening from the domains/genes selector](#opening-from-domains-selector)
    - [Viewing heatmaps in the panel](#heatmaps-panel)
    - [Navigation from annotations](#annotation-navigation)
    - [Other](#other)

<h2 id="background">Background</h2>

NGB shall support heatmaps displaying.
Heatmap is a data visualization technique that shows magnitude of a phenomenon as color in two dimensions.
Under the hood, it is one of the way to display a plain matrix.
In genetic researches, this may be convenient to present gene expression data for different samples, e.g.:
  ![image](https://user-images.githubusercontent.com/45459424/127682721-308fa8af-3568-4110-b070-b258f20eaf0d.png)
In heatmap above, the data is displayed in a grid where each row represents a sample and each column represents a gene. The color and intensity of the boxes is conventionally depicted changes of gene expression.
On the other hand, via heatmap any other data can be represented, main condition that data should be aggregated in plain matrix with two dimensions.

Further, it would be useful to implement also the support of dendrograms for heatmaps.
A dendrogram is a diagram representing a tree.
This diagrammatic representation is frequently used to:
- show the clustering of genes or samples
- display the evolutionary relationships among various biological taxa. In this case, the dendrogram is also called a phylogenetic tree

Dendrogram are placed in the margins of heatmaps (it could be margin of one dimension only or both dimensions at once).
For heatmap above, such dendrogram for the clustering of genes could be:
  ![image](https://user-images.githubusercontent.com/45459424/127683375-35926e1f-004f-4284-8075-d382ed79b8b2.png)

<h2 id="approach">Approach</h2>

<h3 id="heatmap-file-registration">Heatmap file registration</h3>

First of all, we should implement the registration of heatmap files in NGB.

Let's consider that heatmap should be presented by two dimensions matrix (organized as a table in the origin file):
wherein, first row of the table should be perceived as the headers set for columns, first column of the table should be perceived as the headers set for rows, e.g.:
```
          Gene1   Gene2   Gene3
Sample1   0.03    0.04    -0.01
Sample2   -0.3    0.2     0
Sample3   -0.12   0       0.2
```

Possible file formats that should be currently supported for heatmaps:
- `tsv` (content - tab-delimited data)
- `csv` (content - comma-delimited data)

Matrices cells can contain different data - numeric values, string values or be empty.

<h4 id="dendrogram-file-registration">Dendrogram file registration</h4>

Dendrogram file (phylogenetic tree file) is a plain `txt` file that content tree-date in [**_Newick_**](http://wiki.christophchamp.com/index.php?title=Newick_phylogenetic_tree_format) format, e.g.:
```
((Human:0.1,Gorilla:0.1):0.4,(Mouse:0.2,Rat:0.2):0.3);
```

File(s) with the dendrogram for a heatmap should be registered separately:
- this can be performed simultaneously during the heatmap registration (e.g. via some _not-required_ option)
- **_or_** this can be performed after the heatmap file registration
 
Features:
- via the _mandatory_ option, the side to which the dendrogram belongs shall be specified (rows/columns)
- there shall not be the ability to add more than one dendrogram to each side of the heatmap. An attempt to add the second dendrogram to the side that already has one - a new one should "rewrite" the previous dendrogram
- there shall be the ability to remove a dendrogram from a heatmap
- registered dendrogram files shall not be displayed in the datasets tree (at the "Datasets" panel)

<h4 id="rowcolumn-annotation-file-registration">Row/column annotation file registration</h4>

For each row/column in a histogram, user may wish to show an additional info (near the row/column header) - it can be some notes or additional label/mark.

File(s) with row/column annotations for a heatmap should be registered separately:
- this can be performed simultaneously during the heatmap registration (e.g. via some _not-required_ option)
- **_or_** this can be performed after the heatmap file registration

Possible file formats that should be currently supported for annotation files:
- `tsv` (content - tab-delimited data)
- `csv` (content - comma-delimited data)

In this file, there should be only two columns:
- first column contains labels. These labels shall match ones from row and column headers of main heatmap file
- second column contains annotation value – it can be any text

Example:
```
Sample1    Gorilla_male
Gene1      ACKR1
Gene2      MICB
Sample3    Gorilla_female
```
 
Features:
- there shall not be the ability to add more than one annotation file to the heatmap. An attempt to add the second annotation file to the heatmap that already has one - a new one should "rewrite" the previous annotation
- there shall be the ability to remove an annotation from a heatmap
- registered annotation files shall not be displayed in the datasets tree (at the "Datasets" panel)

<h4 id="cell-annotation-file-registration">Cell annotation file registration</h4>

For each cell in a histogram, user may wish to show an additional info (in the cell itself or in a tooltip, e.g.) - it can be some notes or additional label/mark.

File(s) with cell annotations for a heatmap should be registered separately:
- this can be performed simultaneously during the heatmap registration (e.g. via some _not-required_ option)
- **_or_** this can be performed after the heatmap file registration

Possible file formats that should be currently supported for cell annotation files:
- `tsv` (content - tab-delimited data)
- `csv` (content - comma-delimited data)

In this file, there should the matrix of the same size as the original heatmap matrix:
- first row and first column shall contain the same labels as the original heatmap matrix
- cells contain any text values (annotations) – for each cell this data corresponds to a cell from the main heatmap file

Example:
```
          Gene1               Gene2               Gene3
Sample1   Annotation_S1_G1    Annotation_S1_G2     
Sample2                       Annotation_S2_G2    Annotation_S2_G3
Sample3                       Annotation_S3_G2      
```

Features:
- there shall not be the ability to add more than one cell annotation file to the heatmap. An attempt to add the second cell annotation file to the heatmap that already has one - a new one should "rewrite" the previous annotation
- there shall be the ability to remove a cell annotation from a heatmap
- registered cell annotation files shall not be displayed in the datasets tree (at the "Datasets" panel)

<h3 id="heatmap-displaying">Heatmap displaying</h3>

A heatmap is a graphical representation of data that uses a system of color-coding to represent different values.
So, any table uploaded in the way described above should be presented as a rectangular grid, where each cell is highlighted with some color according to the numerical value of the cell.

For mockups below, let's use:
- the following main matrix:
  ![image](https://user-images.githubusercontent.com/45459424/127886170-ef96919e-94d8-4ea1-a20e-1f370c84802a.png)
- the following phylogenetic tree for rows:
  `(((Sample3:0.2,Sample2:0.2):0.2,Sample1:0.4):0.2,Sample4:0.6);`
- the following row/column annotations file:
```
Sample1    Gorilla_male
Gene2      ACKR1
Gene3      MICB
Sample3    Gorilla_female
```
- the following cell annotations file:
```
          Gene1               Gene2               Gene3               Gene4
Sample1   Annotation_S1_G1    Annotation_S1_G2                        Annotation_S1_G4
Sample2                       Annotation_S2_G2    Annotation_S2_G3
Sample3                       Annotation_S3_G2
Sample1   Annotation_S4_G1                                            Annotation_S4_G4
```

Heatmap displaying features:
- heatmap should be displayed as additional track in the "Browser" panel
- heatmap at the track shall not be linked to any genome coordinates, it should be displayed static (i.e. it should not be change or move when user change the scale, move the ruler or change the coordinates, e.g.)
- heatmap image should have a rectangular form and be auto-scaled only to the track size
- each row in the heatmap should be subscribed at the heatmap margin (according to the corresponding row "header")
- each column in the heatmap should be subscribed at the heatmap margin (according to the corresponding column "header")
- near the heatmap, the color legend should be placed

Example (for the table specified above):
  ![image](https://user-images.githubusercontent.com/45459424/132306790-7048f2cf-e23b-48be-b994-43b512af83eb.png)

<h4 id="color-scheme-configuration">Color scheme configuration</h4>

There are two types of heatmap colorizing shall be supported:
- **_continuous_** - in this case, each cell should be colorized according to a gradient color "ruler". There, 3 colors should be configured - for minimum value (from all matrix), for maximum value (from all matrix) and middle value (average between minimum and maximum). Colors for all other values in matrix are selected from a gradient (**_Note_**: this scheme is shown at the example above)
- **_discrete_** - in this case, each cell should be colorized according to a set of discrete values. There, all colors should be configured manually

The color scheme can be configured on the fly, via the special pop-up.
It can be opened by the "Color scheme" control in the header of the heatmap track.

Possible view of the "Color scheme" pop-up:
  ![image](https://user-images.githubusercontent.com/45459424/127899461-a64f933d-f5c2-462e-ac5e-49f7d50f5c21.png)

Here, the following controls should be:
- dropdown list for the color scheme selecting - _continuous_ (by default) or _discrete_
- block of the colors settings according to the selected type
- control to set color for empty cells (without values)

After all colors are configured, user should click the **Save** button. The pop-up will be closed and performed changes shall automatically applied to the heatmap (it will be redrawn).

Color settings of _continuous_ scheme:
- there should be 3 controls for the color setting:
  - "_Low color_" - for color of the minimum value (from all matrix)
  - "_High color_" - for color of the maximum value (from all matrix)
  -  "_Middle color_" - for color of the middle value (average between minimum and maximum)
- other cell colors at the heatmap should be defined based on these 3 colors, gradient between them and cell value

Color settings of _discrete_ scheme:
- by default, there should be a control for a single color setting, e.g.:
  ![image](https://user-images.githubusercontent.com/45459424/127902389-bd693457-db8b-4732-8e90-8f21170f4fa3.png)
  - in such case, all non-empty cells of the table shall be colorized in this color, e.g.:
  ![image](https://user-images.githubusercontent.com/45459424/132307132-f5e15161-a1ca-45ec-a28f-c788a04e80c4.png)
- by the button **Add color** user can add a color:
  - for each color (if colors count equals or larger than 2), an interval of values should be specified for which the specific color should be applied
  - for specifying intervals, the corresponding fields should appear - "_From_" and "_To_" - near each color. Wherein, the number specified in the field "_From_" should be perceived in interval as "greater or equal" and the number in the field "_To_" - should be perceived in interval as "less"
  - for removing a color from the scheme (if colors count equals or larger than 2), the special button should appear near each color
  ![image](https://user-images.githubusercontent.com/45459424/127905137-163b8832-c433-43bd-bc53-bbff74d2a598.png)
- there shall not be the ability to remove all colors (at least one color shall always remain)
- there shall not be the ability to save the scheme where intervals are intersected
- for each color, both values ("_From_" and "_To_") should be specified
- if the cell value is not in any specified interval, it should be colorized in the color for empty cells
- example of the intervals specifying:
  ![image](https://user-images.githubusercontent.com/45459424/127906170-984bdd9b-000a-4c81-8bd4-a2a50252feff.png)
  And the corresponding heatmap:
  ![image](https://user-images.githubusercontent.com/45459424/132307650-bb865983-2230-4eec-86b8-a91d8db05a8e.png)

> **Notes**: there should be some restrictions of the color scheme choice according to the origin heatmap data content:
> - if the matrix contains a string value - at least in 1 cell - for such heatmap:
>    - the _continuous_ scheme should be disabled
>    - in the _discrete_ scheme, for each different value should be automatically set a color, user should have the ability to configure them, e.g.:
>    ![image](https://user-images.githubusercontent.com/45459424/136189956-2f0cfcc9-0f12-42fd-bf50-1649f4650759.png)
> - if the matrix contains only numeric values - for such heatmap, in the _discrete_ scheme:
>    - if there are only several different values (_number of values is below the certain threshold that can be set by the system setting_) - for each value, there should be automatically set the separate color. But user also shall have the ability to set the color for the range of values by the way similar to described above, e.g.:
>    ![image](https://user-images.githubusercontent.com/45459424/136191246-3679c823-caff-42ea-89e0-6a8155330f89.png)
>    - if there are lots of different values (_number of values equals or greater the certain threshold_) - colors are set only for the ranges of values - fully by the way as described in the approach above

<h4 id="displaying-additional-details">Displaying additional details</h4>

If the current heatmap has any additional linked files (_phylogenetic tree_ and/or _row/column annotation_) - the menu **Display** shall be shown in the heatmap track header.  
It should contain only items according to additional heatmap files (if any of them are missing - the corresponding items should be hidden):
  ![image](https://user-images.githubusercontent.com/45459424/132314791-627e4fd7-2263-4912-ace1-60f48f2de67c.png)

If user selects the "Dendrogram" item - the phylogenetic tree(s) should be shown for the heatmap:
- tree for rows should be displayed on the side of the heatmap
- tree for columns should be displayed on the top side of the heatmap
- rows and/or columns of the heatmap should be reordered according to the tree info (from the original tree-file)
![image](https://user-images.githubusercontent.com/45459424/132312525-f78a0616-66e2-4589-accd-98fcbf7f670a.png)

If user selects the "Row/column annotation" item - the corresponding annotation labels should appear on the both axis of the heatmap:
- annotations for rows should be displayed near the rows' headers
- annotations for columns should be displayed under the columns' headers
![image](https://user-images.githubusercontent.com/45459424/132313778-354de727-d33d-4dad-a23d-bb97c179bb22.png)

Both options described above can be enabled simultaneously.

If the current heatmap has an additional linked cell annotation file - the corresponding annotation labels should appear in tooltips when hovering any cell of the heatmap:
![image](https://user-images.githubusercontent.com/45459424/132315935-95b0ccc3-ee06-4818-b700-32c1097b879a.png)

If the current heatmap does not have a cell annotation file - in tooltips when hovering any cell, only corresponding row and column headers should be displayed.

<h3 id="additional-options">Additional options</h3>

<h4 id="displaying-at-summary">Displaying at the Summary page</h4>

If for the dataset containing a heatmap file the **Summary** view is selected and the heatmap file is selected in the **Datasets** panel as well - the heatmap should be shown at the **Summary** view:
- heatmap should be shown below the panels with reference info, datasets list, files list, variants graphs
- heatmap should be shown similarly to the described view at the track with the following details:
  - no header should be shown
  - color scheme should be _continuous_, by default

Example:  
![image](https://user-images.githubusercontent.com/45459424/134949494-92104aae-ab64-4826-acc3-3733f404d6e1.png)

> **_Note_**: if in a dataset several heatmaps are selected:
> - only one should be shown in the **Summary** view
> - there shall be a specific control (near the heatmap files block in the **Summary**) that will allow to select which of heatmaps to show

<h4 id="opening-from-domains-selector">Opening from the domains/genes selector</h4>

There shall be the ability to add a heatmap file as an annotation for a reference.
If the reference has the heatmap annotation file - such heatmap should not be shown as separate file in the **Datasets** panel, but user shall have the ability to open it via the _Domains/Genes_ selector in the **Browser** header, e.g.:
![image](https://user-images.githubusercontent.com/45459424/134953493-eddc1320-23fa-49a5-b978-7defb4dc7d59.png)

If no position at a reference is specified and the **Summary** view is selected - the heatmap should be shown in the way as described above - [Displaying at the Summary page](#displaying-at-summary).
If any position at a reference is specified - the heatmap should be shown as the separate track - as described above [Heatmap displaying](#heatmap-displaying).

> **_Note_**: if the reference has the same annotation heatmap file (and this annotation is enabled) that separately was also added and selected in the dataset:
> - the heatmap panel shall be shown only once in the **Browser** with the _Summary_ view
> - if any reference position is specified, the heatmap tracks shall be shown separately - one for annotation and one for heatmap file itself

<h4 id="heatmaps-panel">Viewing heatmaps in the panel</h4>

A new panel **HEATMAPS** shall be added.
User shall have the ability to enable this panel via the "Views" item of the main menu.
In this panel, heatmaps of the current selected dataset should be shown:
- all heatmaps registered as the current reference annotations
- heatmaps registered as separate files and selected in the **DATASETS** panel

Example:
![image](https://user-images.githubusercontent.com/45459424/136233799-2a6be3bf-da7e-42c1-bb9a-37ff29964034.png)

Features of displaying:
- only one heatmap shall be displayed at the **HEATMAPS** panel at once
- heatmap should be shown similarly to the described view at the track with the following details:
  - no header should be shown
  - color scheme should be _continuous_, by default
- there shall be the dropdown list with heatmaps available for the current selected dataset and its files - user can select any heatmap from the list - it will be shown, multi-select shall not be supported:
  ![image](https://user-images.githubusercontent.com/45459424/136236847-f14b28b1-5216-4256-b7cc-ad8668deb9bb.png)
- colorizing settings shall be available by click the heatmap legend as well
- displaying and color config for any heatmap in the **HEATMAPS** panel shall not affect the displaying or colorizing of the same heatmap in the _Summary_ view or in the track - they all are independent

> **_Notes_**:
> - if no dataset is selected or selected dataset doesn't have any heatmap - the **HEATMAPS** panel shall display the corresponding message
> - if the reference has the same annotation heatmap file that separately was also added and selected in the dataset - this heatmap shall be shown only once in the dropdown list at the **HEATMAPS** panel

<h4 id="annotation-navigation">Navigation from annotations</h4>

All heatmaps' _labels_ (row/column headers, row/column annotations, cell annotations) shall support "link" format for their content as well - that users can have the ability for the fast navigation through other NGB objects from a heatmap.
I.e. user shall have the ability to click a heatmap label - and the certain object (according to the label link type) shall be opened in NGB.

Whether the set of the corresponding heatmap labels are links should be defined during the heatmap file or heatmap annotation file CLI registration - via the additional options.
Format of the additional option shall be like: `-<entity_type> <link_type>`
Where:
- **<entity_type>** - the option name that should define the certain heatmap entity, which set of labels shall be perceived as links
- **<link_type>** - the type of NGB objects to which the link should provide

The following heatmap **entity types** options shall be supported:
- `-col-at` (`--heatmap-col-at`) - if that option is set, column headers and column annotations shall be perceived as links. In this case - in first order, the column annotation should be used as link, if it is absent - the column header should be used as link.
- `-row-at` (`--heatmap-row-at`) - if that option is set, row headers and row annotations shall be perceived as links. In this case - in first order, the row annotation should be used as link, if it is absent - the row header should be used as link.
- `-cell-at` (`--heatmap-cell-at`) - if that option is set, cell annotations shall be perceived as links

The following **link types** shall be supported:
- `COORDINATE` - heatmap label is perceived as coordinates at the reference. When user clicks such label - the corresponding part of reference shall be opened
- `GENE` - heatmap label is perceived as gene name. When user clicks such label - the search over genes shall be performed by the label name and then, the result at the GENES track shall be opened
- `REFERENCE` - heatmap label is perceived as reference name. When user clicks such label - the search over loaded references shall be performed by the label name and then, the first found dataset with the such reference shall be selected and opened
- `DATASET` - heatmap label is perceived as dataset name. When user clicks such label - the search over datasets shall be performed by the label name and then, the found dataset shall be selected and opened

Label names should be perceived as is for the corresponding searches.
The following format of the label text should be expected:
- for the `COORDINATE` link type:
   - `<reference_name>:<chromosome_name>:<start_position>-<end_position>` - to navigate to the specified reference and the chromosome and the range on it
   - `<reference_name>:<chromosome_name>:<position>` - to navigate to the specified reference and the chromosome and the position on it
   - `<chromosome_name>:<start_position>-<end_position>` - to navigate to the specified chromosome and the range on it of the current opened reference
   - `<chromosome_name>:<position>` - to navigate to the specified chromosome and the position on it of the current opened reference
- for the `GENE` link type:
   - `<reference_name>:<gene_name>` - to navigate to the specified reference and the gene on it
   - `<gene_name>` - to navigate to the specified gene on the current opened reference
- for the `REFERENCE` link type:
   - `<reference_name>` - to navigate to the specified reference
- for the `DATASET` link type:
   - `<dataset_name>` - to navigate to the specified dataset

> **_Notes_**:
> - if the position to navigate is not found - nothing shall be performed after the link click
> - for cell annotations, following the link should be performed by the cell click

<h4 id="other">Other</h4>

**Saving in session**
Opened heatmaps should being saved in sessions. I.e. if user opens a heatmap (e.g. in the **HEATMAP** panel) and save the session - then, if such session will be opened later - the corresponding heatmap should be opened.
If any session was saved without any opened heatmap - then, if such session will be opened later - no heatmaps should be displayed in the **HEATMAP** panel even if any heatmap was shown there before this session opening.

**Saving the view of the last opened heatmap**
As from the heatmap of one reference can be opened other reference (due to described [navigation](#annotation-navigation)) - in such case, the last opened heatmap should remain in the **HEATMAP** panel (should not be closed automatically).

**Comment by Tatyana2022** [comment]
**BUG was found:**
The Heatmap file is dowloaded in xhtml format

**Steps to reproduce:**
1. Go to NGB
2. Go to DATASETS panel
3. Go to **Halo_pallet_skmer** file in **heatmap** dataset
4. Right click on file
5. Click **Download file**
6. Look at downloaded file

**Expected result:**
**Halo_pallet_skmer** file should be downloaded in .tsv format

**Actual result:** 
**Halo_pallet_skmer** file is downloaded as **download.xhtml** 

**Extra Details:**
The same behavior almost for all heatmaps files
The Heatmap file is correctly downloaded from the dataset and from track. 

@rodichenko @mzueva @okolesn 

***Verified***

**Comment by Tatyana2022** [comment]
**BUG was found:**
The number of openable  **Color scheme** windows depends on the number of heatmaps opened on the panel

**Steps to reproduce:**
1. Go to NGB
2. Go to DATASETS panel
3. Select **SV_Sample1** dataset
4. Open **Heatmap** panel in **VIEWS** menu
5. Expand **Source** dropdown field
6. Select any other heatmap (e.g. **heatmap-ap**)
7. Open **Color scheme** window

**Expected result:**
 One **Color scheme** window should be opened

**Actual result:** 
2 **Color scheme** window is opened. 

**Extra Details:**
1) The number of open **Color scheme** windows depends on the number of heatmaps that were opened in this panel. 
2) New window opens for a specific display when switching between displays (e.g. switching from Discrete to Continuous). The background behind the windows turns black
<img src="https://user-images.githubusercontent.com/83421565/135620880-3ac4af38-c012-4373-beae-fb29acef2961.png" width="200" />

@rodichenko @mzueva 

***Verified***

**Comment by rodichenko** [comment]
> **BUG was found:** The number of openable **Color scheme** windows depends on the number of heatmaps opened on the panel

@Tatyana2022 should be fixed by 12b9c9eefb941f089df4ec56d53f6f497ebeab51

***Verified***

**Comment by Tatyana2022** [comment]
**Bug was found:**
All selected heatmaps can be displayed in the Source field at the same time


**Steps to reproduce:**
1. Go to NGB
2. Select **SV_Sample1** dataset in DATASETS panel
3. Open HEATMAP panel in **VIEWS** menu
4. Select **Numeric** heatmap in **Source** dropdown menu
5. Select **test** heatmap
6. Select **Hallo_pallet_skmr** heatmap
7. Look at **Source** field

**Expected result:**
The last open heatmap should be displayed in the **Source** field: **Hallo_pallet_skmr** 

**Actual result:**
All opened heatmaps are displayed in the **Source** field: **Hallo_pallet_skmr, Numeric, test**
![image](https://user-images.githubusercontent.com/83421565/135805709-44f7b2a1-43d2-4284-993f-e98e80e10811.png)

@rodichenko @mzueva 

***Verified***

**Comment by Tatyana2022** [comment]
**Bug was found:**
**Source** field becomes empty after applying a page refresh

**Steps to reproduce:**
1. Go to NGB
2. Select **SV_Sample1** dataset in DATASETS panel
3. Open HEATMAP panel in **VIEWS** menu
4. Reload the page (F5)
5. Look at **Source** field

**Expected result:**
**Source** field should be filled by **Hallo_pallet_skmr** heatmap

**Actual result:**
Empty field displays in the **Source** field 
![image](https://user-images.githubusercontent.com/83421565/135809903-b2d893a3-baf4-4330-9cf0-ff97d7881158.png)


@rodichenko @mzueva 
***Verified***

**Comment by Tatyana2022** [comment]
**Bug was found:**
Heatmap in csv format is not added to the NGB

Input data: 
File name - New.csv

              Gene1   Gene2   Gene3
      Sample1 0.03    0.04    -0.01
      Sample2 -0.3    0.2     0
      Sample3 -0.12   0	0.2

Response in swagger: 
![image](https://user-images.githubusercontent.com/83421565/135815921-eac94a25-b4a6-47b6-b3d1-89925ff9add7.png)

**Extra details:**
After that, a file with the name New.tsv was added to the application without errors. Only the format name has been changed

@mzueva

**Comment by Tatyana2022** [comment]
**BUG was found:**
Heatmaps in .csv, .tsv formats are not registered in the NGB

![image](https://user-images.githubusercontent.com/83421565/135846286-21c88aeb-d34a-4f11-b0ed-5bd42b2f18bc.png)

@mzueva

**Comment by Tatyana2022** [comment]
**Bug was found:**
Annotations do not appear on row and column names

**Prerequisites:**
Heatmap file: New.tsv
          Gene1   Gene2   Gene3
Sample1   0.03    0.04    -0.01
Sample2   -0.3    0.2     0
Sample3   -0.12   0       0.2

Annotation file: Annotation.tsv
 Gene1 ACKR1
 Gene2 MICB
 Sample1 Gorilla_male

**Steps to reproduce:**
1. Go to NGB
2. Go to **SV_Sample1** dataset
3. Select **New** file (second in order)
4. Open **HEATMAPS** in **VIEWS** menu
5. Select **New1** in **Source** dropdown field
6. Look at names row and column names

**Expected result:** 
The following values should be displayed:
column names - Gene1 ACKR1,  Gene2  MICB, Gene3
row names - Sample1    Gorilla_male, Sample2, Sample3

**Actual result:**
No annotations for row and column names: 
column names: Gene1,   Gene2 ,  Gene3
row names: Sample1, Sample2, Sample3

<img src="https://user-images.githubusercontent.com/83421565/135854188-a43355df-5b47-46dc-87a7-b395f1b20583.png" width="200" /> <img src="https://user-images.githubusercontent.com/83421565/135855288-48a59fba-c595-4f94-aa80-af31b920ee47.png" width="200" /> <img src="https://user-images.githubusercontent.com/83421565/135855549-8fd7da97-bca3-4f31-97f7-cc1a47a9279a.png" width="200" />

@rodichenko @okolesn @mzueva 

***Verified***

**Comment by Tatyana2022** [comment]
**Bug was found:**
Incorrect display of data in cells for exponential notation (1E+x)

**Steps to reproduce:**
1. Go to NGB
2. Go to **SV_Sample1** dataset in **DATASETS** panel
3. Select **TSV-Heatmap** file
4. Open **HEATMAP** panel from **VIEWS** menu
5. Expand **Source** field
6. Select **TSV-Heatmap** value
7. Go to the table with colored cells
8. Hover over the cell at the intersection of the ** test.1 ** and ** 1 ** headers

**Expected result:**
The tooltip should display the following value: 1E+17

**Actual result:**
The following value display in the tooltip: **100000000000000000**
<img src="https://user-images.githubusercontent.com/83421565/136022246-ffaaa611-341d-4ad7-82b7-8b66ab878efe.png" width="200" />
<img src="https://user-images.githubusercontent.com/83421565/136022646-95f540e6-c588-44e6-ae75-fe3025c9985f.png" width="200" />

**Extra Details:**
Since the values differ in the scale and on the table, the set color is not pulled into the cell

@rodichenko @mzueva 

***Verified***

**Comment by Tatyana2022** [comment]
**Bug was found:**
Empty values in cells are displayed as 0

**Steps to reproduce:**
1. Go to NGB
2. Go to **SV_Sample1** dataset in **DATASETS** panel
3. Select **CSV-Heatmap** file
4. Open **HEATMAP** panel from **VIEWS** menu
5. Expand **Source** field
6. Select **CSV-Heatmap** value
7. Go to the table with colored cells
8. Hover over the cell at the intersection of the **Gene3** and **Sample1** headers

**Expected result:**
**NULL** value should be displayed in the tooltip of cell

**Actual result:**
**0** value display in the tooltip of cell

**Extra details:**
In the color settings, the empty value is assigned a color

@rodichenko @mzueva 

***Verified***

**Comment by Tatyana2022** [comment]
**Bug was found:**
An empty value has string type by default

**Prerequisites:**
Heatmap file: **CSV-Heatmap**
, Gene1, Gene2, Gene3, Gene4
Sample1, -1, -1.2, , 1.4
Sample2, 0.5, 0.2, -0.7, 100
Sample3, -1.8, 1.9, -0.1, 1
Sample4, 2.3, 1.1, -1.3, 0.6


**Steps to reproduce:**
1. Go to NGB
2. Go to **SV_Sample1** dataset in **DATASETS** panel
3. Select **CSV-Heatmap** file
4. Open **HEATMAP** panel from **VIEWS** menu
5. Expand **Source** field
6. Select **CSV-Heatmap** value
7. Сlick on the color bar on the right
8. Expand **Type**field

**Expected result:**
**Discrete** and **Continuous** types should be able for choice

**Actual Result:**
Only **Discrete** type is active in type field
![image](https://user-images.githubusercontent.com/83421565/136078677-fdc1c6ff-205c-49df-9d88-5c0dae32ad3f.png)

@okolesn @mzueva 

***Verified***

**Comment by Tatyana2022** [comment]
**Bug was found:**
Incorrect display of values for a heatmap with a large number of cells in the Color scheme window

**Steps to reproduce:**
1. Go to NGB
2. Go to **SV_Sample1** dataset in **DATASETS** panel
3. Select **Halo_pallet_skmer** file
4. Open **HEATMAP** panel from **VIEWS** menu
5. Expand **Source** field
6. Select **Halo_pallet_skmer** value
7. Сlick on the color bar on the right
8. Select **Discrete** type
9. Click **Add color** button
10. Look at **From**,**To** values

**Expected result:**
One range field with the minimum and maximum values of the current heatmap should be displayed in the Heatmap Color Scheme (e.g. From 0 To 0.002)

**Actual result:**
All fields are auto-filled with values in ascending order with values, starting from 0 (e.g. From = 0 To =1)
![image](https://user-images.githubusercontent.com/83421565/136163914-77b02d09-8b04-459d-bd28-b7ac73e80e8d.png)

@rodichenko @mzueva

**Comment by Tatyana2022** [comment]
**Nice to have:**
It would be great to make a dividing line between annotation and dataset heatmaps in **Source** field. 
And apply alphabetical sorting separately for annotated heatmaps and for heatmaps from datasets.

@rodichenko @mzueva @NShaforostov 

**Fixed and verified**

**Comment by Tatyana2022** [comment]
**Bug was found:**
CSV cell annotations are not added to the heatmaps

**Steps to reproduce:** 
1. Сreate a  csv file in terminal (e.g. CSV-Annotation1.csv)
2. Fill file with data: 
,Gene1,Gene2,Gene3,Gene4
Sample1,0,1,2,3
Sample2,10,11,12,13
Sample3,20,21,22,23
Sample4,30,31,32,33
3. Add  cell-annotation file to heatmap: **CSV-Annotation1.csv** to **CSV-Heatmap.csv**

**Expected result:**
Cell-annotation data should be added to **CSV-Heatmap.csv** heatmap

**Actual result:**
Cell-annotation data aren't added to the heatmap. An error display in the response: **Incorrect file format**
<img src="https://user-images.githubusercontent.com/83421565/136196975-2534567a-5a91-4959-bf36-868bcf3f9d15.png" width="200" />

@okolesn @mzueva 

***Verified***

**Comment by Tatyana2022** [comment]
**Bug was found:**
Incorrect display of data in the table and an error is displayed when adding a csv cell-annotation to a tsv heatmap

**Prerequisites:**
Heatmap-file: **1_2.tsv**
```
      gene1
sample1 1.1
sample2 1
```

Cell-annotation file: **1_2annotation.csv**
```
,gene1
sample1,Annotation1
sample2,10
```
**Steps to reproduce:**
1. Go to NGB
2. Go to **SV_Sample1** dataset in **DATASETS** panel
3. Select **1_2** file
4. Open **HEATMAP** panel from **VIEWS** menu
5. Expand **Source** field
6. Select **1_2** value
7. Look at table

**Expected result:**
1 by 2 table should be displayed in the panel. 2 cells in this table should be filled with 2 different colors. 
When you hover over a cell, a tooltip with value data and annotations should be displayed

**Actual result:**
2 cells in the table is filled by white color as in "Missing Data" field. No tooltips when hovering over cells.
An error display in the Network tab of Dev-Console
<img src="https://user-images.githubusercontent.com/83421565/136204811-ea1b6547-c41d-4bf6-be5f-0a8e80176fb9.png" width="200" />

@okolesn @rodichenko @mzueva 

***Verified***

**Comment by Tatyana2022** [comment]
**Bug was found:**
It is impossible to add empty values to heatmap files

It should be possible to add the following values as empty fields to the tsv,csv files: missing, spaces, point

@okolesn @mzueva 

***Verified***

**Comment by Tatyana2022** [comment]
**Bug was found:**
Big empty space is displayed between the heatmap block and all others in the Browser window

**Steps to reproduce:**
1. Go to NGB
2. Select **heatmap** dataset in **DATASETS** panel
3. Go to **BROWSER** panel

**Expected result:**
Block of heatmaps should be displayed below all blocks without much empty space

**Actual result:**
Big empty space is displayed between the heatmap block and all others in the Browser window

<img src="https://user-images.githubusercontent.com/83421565/136224993-5409ebf2-05e0-471e-89fa-b6ba0239c296.png" width="200" />

@rodichenko @mzueva 

***Verified***

**Comment by Tatyana2022** [comment]
**Nice to have:**
It would be great to keep the changes applied to the color scheme of the particular heatmap.

@rodichenko @mzueva 

***Verified***

**Comment by Tatyana2022** [comment]
**Bug was found:**
Long row/column headers with tilted (with or without annotations) - lead to header font rasterization in the heatmap

**Prerequisites:**
Heatmap-file: **Long-headers.tsv**
```
        Gorilla_male09876554    Hyperolius_sapiens123456789     ACKR1_male
MICB_female98989888     1	2	3
Gorilla_female123	4	5	6
Abies homolepis 7	8	9
```
**Steps to reproduce:**
1. Go to NGB
2. Go to **SV_Sample1** dataset in **DATASETS** panel
3. Select **Long-headers** file
4. Open **HEATMAP** panel from **VIEWS** menu
5. Expand **Source** field
6. Select **Long-headers** heatmap
7. Look at table

**Expected result:**
The font of displaying row/column headers should be without  rasterization on any length of name 

**Actual result:**
The font of headers are rasterized for names with tilt.
<img src="https://user-images.githubusercontent.com/83421565/136385152-6e74f22a-fda9-45d6-8316-121bbcef4bc4.png" width="200" />

@rodichenko @mzueva 

***Verified***

**Comment by Tatyana2022** [comment]
**Bug was found:**
The error is displayed in a discrete view of the color scheme after switching between heatmaps

**To reproduce:**
1. Go to NGB
2. Select **SV_Sample1** in **DATASETS** panel
3. Go to **Heatmaps** panel
4. Select **Long-headers** heatmap in **Source** field
5. Open **Color Scheme** window
6. Select **Discrete** view in **Type** field (if not selected)
7. Delete all colors, except any one
8. Click **Add color** button
9. Set checkbox in **Range** (if not set)
10. Set any values in the range fields (e.g. 
         1) From 1 To 2 
         2) From 2 To 3
11. Click **Save** button
12.  Select **Test** heatmap in **Source** field
13. Repeat 4-5 steps

**Expected result:**
Previously entered values should be displayed in the range fields. No errors in the window and console. **Save** button should be enabled

**Actual result:**
An error display below range fields in **Color Scheme** window.  No values in the range fields. **Save** button is disabled. 
<img src="https://user-images.githubusercontent.com/83421565/137282754-e7a92a70-525c-4742-842d-574bdb2653eb.png" width="200" />

**Extra Details:**
Errors display in the console
<img src="https://user-images.githubusercontent.com/83421565/137282801-57847f84-9d98-4df0-be8d-b418978d4c3c.png" width="200" />

@rodichenko @mzueva 
***Verified***

**Comment by Tatyana2022** [comment]
**Bug was found:**
The whole heatmap table is colored in "Missing data" color if there is a Null value in the cell

**To reproduce:**
1. Go to NGB
2. Select **SV_Sample1** in **DATASETS** panel
3. Go to **Heatmaps** panel
4. Select **Empty-cells** heatmap in **Source** field
5. Open **Color Scheme** window
6. Select **Discrete** view in **Type** field (if not selected)
7. Click **Save** button
8. Look at Heatmap

**Expected result:**
All cells should be colored according to the specified colors for values. 
The **Null** value should be colored in accordance with the color specified in the **Missing data** button

**Actual result:**
All cells are painted in the color specified in the **Missing data** button. 
**Renedering...** text display in the middle of heatmap table. 
An error display in the console:
```TypeError: Cannot read properties of null (reading 'toString')
    at Gradient.value (app.bundle.js:285)
    at GradientCollection.value (app.bundle.js:286)
    at GradientCollection.value (app.bundle.js:286)
    at ColorScheme.value (app.bundle.js:276)
    at HoveredElement.value (app.bundle.js:686)
    at app.bundle.js:686
    at Array.forEach (<anonymous>)
    at HoveredRenderer.value (app.bundle.js:686)
    at HeatmapDataRenderer.value (app.bundle.js:684)
    at HeatmapDataRenderer.value (app.bundle.js:684)
    at HeatmapView.value (app.bundle.js:683)
    at HeatmapView.value (app.bundle.js:683)
    at HeatmapView.value (app.bundle.js:683)
    at HeatmapView.value (app.bundle.js:683)
    at Heatmap.value (app.bundle.js:275)
    at Heatmap.value (app.bundle.js:275) 
```
<img src="https://user-images.githubusercontent.com/83421565/137287779-a79fa6c1-3194-411a-b612-2dd7d6032ecf.png" width="200" />

**Extra Details:**
When re-saving data in the color scheme - the color scheme is shifted to the right beyond the screen scale in the heatmap panel
<img src="https://user-images.githubusercontent.com/83421565/137288191-8772c423-c15a-4bb4-94a6-cd955680a2e6.png" width="200" />


@rodichenko @mzueva 

***Verified***

**Comment by Tatyana2022** [comment]
**Bug was found:**
Comma display in the Source field after reloading the page

**To reproduce:**
1. Go to NGB
2. Go to **SV_Sample1** dataset in **DATASETS** panel
3. Select **Test** file
4. Go to **Heatmaps** panel
5. Reload the page (F5)
6. Look at **Source** field

**Expected result:**
Page is reloaded without any changes. Name of heatmap display in the **Source** field (Test)

**Actual result:**
A comma appears in front of the heatmap name for a few seconds, then disappears (, Test)

**Extra Details:**
***The same behavior on Production***

@rodichenko @mzueva @DmitriiKrasnov 

***Verified***

**Comment by Tatyana2022** [comment]
**Bug was found:**
The drop-down list of the "Source" field moves after clicking the **maxinise/minimise** button.

**To reproduce:**
1. Go to NGB
2. Go to **SV_Sample1** dataset in **DATASETS** panel
3. Select **Test** file
4. Go to **Heatmaps** panel
5. Click on **maximise** button
6. Expand **Source** drop-down list
7. Click **minimise** button
8.  Expand **Source** drop-down list

**Expected result:**
The drop-down list should be expanded to the width of the field

**Actual result:**
The drop-down list expands to the left of the **Source** field width

https://user-images.githubusercontent.com/83421565/138105138-5809e975-67a0-425e-aeba-75b49f26ddd3.mov

**Extra Details:**
***The same behavior on Production and for all dropdown fields of all panels***

@rodichenko @mzueva @DmitriiKrasnov

**Comment by Tatyana2022** [comment]
**Bug was found:**
The heatmap is not colored and loader text is displayed instead

**To reproduce:**
1. Go to NGB
2. Go to **SV_Sample1** dataset in **DATASETS** panel
3. Select any file (e.g.**Test**)
4. Go to **Heatmaps** panel


**Expected result:**
The heatmap is colored according to the color scheme

**Actual result:**
The heatmap is not colored, the Loader text is displayed instead. 
An error display in the console: ```app.bundle.js:277 Error fetching data: invalid stream header: 0947656E```

<img src="https://user-images.githubusercontent.com/83421565/138241439-1333906b-0fa2-41bc-8044-cdac9ad40c8d.png" width="200" />

**Extra Details:**
The same behavior for all heatmaps and for display on tracks

 @mzueva @okolesn 

***Verified***

**Comment by Tatyana2022** [comment]
**Bug was found:**
**Source** field is empty for last opened heatmap

**To reproduce:**
1. Go to NGB
2. Go to **SV_Sample1** dataset in **DATASETS** panel
3. Select any file (e.g. **Test**)
4. Go to **HEATMAP** panel
5. Relad the page (F5)
5. Uncheck chosen file in DATASETS panel (e.g. **Test**)
6. Go back to **HEATMAP** panel
7. Reload the page
7. Look at **Source** field

**Expected result:**
**Source** field should be filled by name of last opened heatmap (e.g. **Test**)

**Actual result:**
**Source** field is empty

**Extra details:** 
The same behavior when switching between datasets or references

<img src="https://user-images.githubusercontent.com/83421565/138282345-d3cf75ba-62bb-4750-b8f9-f2fca08ca9b8.png" width="200" />


@rodichenko @mzueva 

***Verified***

**Comment by Tatyana2022** [comment]
**Bug was found:**
Errors display in the console after opening a session with heatmaps

**Prerequisites:**
Open any heatmap file

**To reproduce:**
1. Go to NGB
2. Open **Sessions** panel in **VIEWS** menu
3. Go to **Sessions** panel
3. Filter table by **heatmap** value in the **Name** column
4. Click on any session (e.g. Heatmap Test)
5. Look at console

**Expected result:**
Errors shouldn't be displayed in the console

**Actual result:**
Errors display in the console

<img src="https://user-images.githubusercontent.com/83421565/138284738-121a536c-6b35-447c-9e15-129e688f2c71.png" width="200" />


@rodichenko @mzueva 

***Veified***

**Comment by Tatyana2022** [comment]
**Bug was found:**
Missing data are not highlighted in the heatmap

**To reproduce:**
1. Go to NGB
2. Expand **SV_Sample1** in **DATASETS** panel
3. Select **CSVTest2** file
4. Go to **HEATMAP** panel
5. Navigate to cell with **null** value in colored heatmap 

**Expected result:**
The Null value should be colored in accordance with the color specified in the Missing data button (e.g. white)

**Actual result:**
Missing data are highlighted in color specified for non-missing data in the heatmap (e.g. green)

@mzueva @rodichenko 

***Verified***

**Comment by rodichenko** [comment]
> **Bug was found:** Missing data are not highlighted in the heatmap
> 
> @mzueva @rodichenko

@Tatyana2022 should be fixed by c021802857bcb169dc64961b2a1c68cb95f2c1a4

**Comment by rodichenko** [comment]
> **Bug was found:** Errors display in the console after opening a session with heatmaps
> 
> @rodichenko @mzueva

@Tatyana2022 should be fixed by 7edb738c044919ca55e4e5a1ae56150a7dd23cfd

**Comment by rodichenko** [comment]
> **Bug was found:** The heatmap is not colored and loader text is displayed instead
> 
> @mzueva @okolesn

@Tatyana2022 should be fixed by b67e37224b95bdecf4556122ce1b923251a4a92e

**Comment by rodichenko** [comment]
> **Bug was found:** **Source** field is empty for last opened heatmap
> 
> @rodichenko @mzueva

@Tatyana2022 should be fixed by f0955965a0f932b85057efb76437bea085296f8f

**Comment by NShaforostov** [comment]
Docs were added via #631 and located [here](https://github.com/epam/NGB/blob/develop/docs/md/user-guide/tracks-heatmap.md#heatmap-track).

**Comment by Tatyana2022** [comment]
**Bug was found:**
Session opens the first heatmap in the Source list instead of the one set for saving

**To reproduce:**
1. Go to NGB
2. Go to **DATASETS** panel
3. Select all heatmap files (e.g. in **heatmap**  dataset)
4. Go to **HEATMAP** panel
5. Select any non-first heatmap in **Source** field (e.g. **large**)
6. Select any chromosome in **Browser** panel
7. Click on **location** button in the left vertical menu
8. Set Name (e.g. Heatmap large)
9. Click **Save**button
10. Open **Sessions** panel
11. Click on last saved session (e.g. Heatmap large)
12. Look at **Heatmap** panel

**Expected result:**
Saved state should open. Selected heatmap should be displayed (e.g. large)

**Actual result:**
The first heatmap in the **Source** list is displayed in the Heatmap panel (e.g. bamy)

@rodichenko @DmitriiKrasnov @mzueva 

***Verified***

**Comment by Tatyana2022** [comment]
**Enhancement**
Shouldn't be possible to add more than 2 children to the parent in trees 

@okolesn

**Comment by Tatyana2022** [comment]
**Bug was found:**
The cursor is displayed in the color selection box when quickly remove colors in a color scheme

**To reproduce:**
1. Go to NGB
2. Go to DATASETS panel
3. Select **CSVTest** file in **SV_Sample1** dataset
4. Go to HEATMAP panel
5. Open **Color Scheme** window
6. Select **Discrete** view 
7. Remove fields with colors quickly

**Expected result:**
All fields are removed except one. A blue vertical line was not displayed on any rectangle with color during fields deletion

**Actual result:**
Vertical blue line appears on one of the color rectangles when removing colors

<img src="https://user-images.githubusercontent.com/83421565/138869216-9c5aae33-cbd7-4b84-9eff-c5f03dc706ac.png" width="200" />

@rodichenko @DmitriiKrasnov @mzueva

**Comment by Tatyana2022** [comment]
**Bug was found:**
The checkbox is automatically set in the heatmap section of the browser panel

**To reproduce:**
1. Go to **DATASETS** panel
2. Select **large** file in **heatmap** dataset
3. Go to **Browser** panel
4. Set checkbox in **large** value in **HEATMAP** section
5. Go to **DATASETS** panel
6. Unset checkbox for **large** file
7.  Set checkbox for **large** file
8. Look at **HEATMAP** section in **BROWSER** panel

**Expected result:**
The checkbox should be unchecked for the **large** file in **HEATMAP** section. 
**large** heatmap shouldn't be displayed in the **BROWSER** panel

**Actual result:**
Checkbox is set for **large** heatmap. 

**Extra details:**
 Checkbox is set for **large** heatmap and heatmap is displayed in the **BROWSER** panel
If the checkbox is unchecked, then the heatmap will appear
If checkbox is set again, then an error display in the console:
`app.bundle.js:690 Uncaught TypeError: Cannot read properties of undefined (reading 'size')
    at RowsRenderer.value (app.bundle.js:690)
    at RowsRenderer.value (app.bundle.js:690)
    at app.bundle.js:690`

https://user-images.githubusercontent.com/83421565/138927811-39e50fa4-a898-4808-b6ee-a8a99a6d141a.mov

@rodichenko @DmitriiKrasnov @mzueva 

***Verified***

**Comment by Tatyana2022** [comment]
**Bug was found:**
Incorrect display of data in color scheme for exponential notation (1E+x)

**Steps to reproduce:**
1. Go to NGB
2. Go to **SV_Sample1** dataset in **DATASETS** panel
3. Select **Empty4.csv** file
4. Open **HEATMAP** panel from **VIEWS** menu
5. Expand **Source** field
6. Select **Empty4.csv** value
7. Open **Color Scheme** window
8. Select **Discrete** view in expanded menu
9. Look at values in fields

**Expected result:**
**1E+5** value should be displayed in one of the field

**Actual result:**
**100000** value display in the **Color Scheme** instead 1E+5 
<img src="https://user-images.githubusercontent.com/83421565/139202384-3453d5a0-91fa-486b-9963-a4376c64bee6.png" width="200" />

**Extra details:**
The same behavior for **Testing1** heatmap

@rodichenko @mzueva @DmitriiKrasnov @okolesn

**Comment by Tatyana2022** [comment]
**Bug was found:**
An errors are displayed for negative numbers when reopening heatmap and color scheme

**To reproduce:**
1. Go to NGB
2. Go to **DATASETS** panel
3. Select **Heatmap_CSV** file in **heatmap** dataset
4. Go to **BROWSER** panel
5. Open **Heatmap_CSV** in **Heatmap** visual block
6. Open **Color Scheme**
7. Select **Discrete** view (if not selected)
8. Click **Save** button
9. Close **Heatmap_CSV** in **Browser** panel
10. Repeat 5-6 steps
11. Look at fields in the **Color Scheme**

**Expected result:**
All existing values of heatmap are displayed in the window without errors

**Actual result:**
An errors display under negative values: **This configuration conflicts with another one**
<img src="https://user-images.githubusercontent.com/83421565/139231046-a10236c7-3647-4443-a424-6a4dba635bda.png" width="200" />

@rodichenko @DmitriiKrasnov @mzueva 

**Fixed and verified**

**Comment by Tatyana2022** [comment]
**Bug was found:**
The values for the first row of the cell-annotation are duplicated for all subsequent rows of the heatmap

**Preconditions:**
Cell-annotation file: **Cell-AnnotationCSV.csv**
``` 
Gene,Sample1,Sample2,Sample3,Sample4,Sample5
BRCA1,BRCA1,Fruitfly,GRCH38,chr2:29224756-29224807,BRCA5
BRCA2,BRCA2-001,BRCA2-002,BRCA2-003,BRCA2-004,BRCA2-005
KRAS,kras-001,kras-002,kras-003,kras-004,kras-005
ROS1,Ros1,Ros2,Ros3,Ros4,Ros5
```

**To reproduce:**
1. Go to NGB
2. Go to **DATASETS** panel
3. Select **Heatmap_CSV** file in **heatmap** dataset
4. Go to **BROWSER** panel 
5. Set checkbox in **Heatmap_CSV** file
6. Navigate to cells
7. Compare values from preconditions with values in tooltips

**Expected result:**
The values of cell-annotations in tooltips and their order should correspond to the values from the precondition

**Actual result:**
Only the values from the first row of the heatmap correspond to the values from the precondition, in subsequent rows duplicates of the first row are displayed:
```
Gene,Sample1,Sample2,Sample3,Sample4,Sample5
BRCA1,BRCA1,Fruitfly,GRCH38,chr2:29224756-29224807,BRCA5
BRCA2BRCA1,Fruitfly,GRCH38,chr2:29224756-29224807,BRCA5
KRAS,BRCA1,Fruitfly,GRCH38,chr2:29224756-29224807,BRCA5
ROS1,BRCA1,Fruitfly,GRCH38,chr2:29224756-29224807,BRCA5
```
<img src="https://user-images.githubusercontent.com/83421565/139387666-ebd604ea-9e98-451a-820e-778d3428b8a7.png" width="200" />
@mzueva @okolesn

**Comment by Tatyana2022** [comment]
**Bug was found:**
Cells that are not part of the tree structure are colored as missing data in the heatmap

**Preconditions:**
Cell-annotation file: **Tree.txt**
``` 
(((Sample1:0.2,Sample2:0.2):0.2,Sample3:0.4):0.2,Sample4:0.6);
```

**To reproduce:**
1. Go to NGB
2. Go to **DATASETS** panel
3. Select **Heatmap_CSV** file in **heatmap** dataset
4. Go to **BROWSER** panel 
5. Set checkbox in **Heatmap_CSV** file
6. Look at heatmap

**Expected result:**
All cells should be colored according to the heatmap regardless of the tree

**Actual result:**
Сells not included in the tree structure - colored as Missing data

<img src="https://user-images.githubusercontent.com/83421565/139398508-72157d25-7d51-490e-b3c0-86a5c67e9fef.png" width="200" />

**Extra Details:**
Annotations also do not work for columns not included in the tree

@mzueva @rodichenko @DmitriiKrasnov

**Comment by Tatyana2022** [comment]
**Bug was found:**
Multi-select for heatmaps available in the Browser panel

**To reproduce:**
1. Go to NGB
2. Go to **DATASETS** panel
3. Go to **heatmap** dataset
4. Select **tree** and **large** files
5. Go to **BROWSER** panel
6. Select **tree** file
7. Select **large** file
8. Look at heatmap visual block in browser panel

**Expected result:**
The checkbox should be set for the last selected heatmap only(for **large**), and for the previous one it should be unchecked (for **tree**)

**Actual result:**
All heatmaps are selected
<img src="https://user-images.githubusercontent.com/83421565/143867680-12400e23-87e6-427d-b4cb-19c26cdd55ad.png" width="200" />

@mzueva @rodichenko 
***Verified***

**Comment by Tatyana2022** [comment]
**Bug was found:**
The heatmap with text-cells is not colored according to the Color Scheme

**To reproduce:**
1. Go to NGB
2. Go to **DATASETS** panel
3. Go to **heatmap** dataset
4. Select **Text_cells(non-empty)** files
5. Go to **BROWSER** panel 
6. Set checkbox in **Text_cells(non-empty)** heatmap
7. Open **Color Scheme** window
7. Remember colors added to existing values in the table
8. Close **Color Scheme**
9. Compare the colors of the painted cells in the table with the colors in the Color Scheme

**Expected result:**
All cells should be painted in colors according to the Color Scheme. 

**Actual result:**
The whole table is colored red. 
Сell names are not indicated in the color scale to the right of the table. 
**Data** name displays for the last color
<img src="https://user-images.githubusercontent.com/83421565/144012521-c676c4a2-a68f-4821-aa3f-70290a91bcb0.png" width="200" />
<img src="https://user-images.githubusercontent.com/83421565/144012653-9bcc87e7-1c6d-4585-ac4c-96e0b457df3d.png" width="200" />

**Extra details:**
The same behavior for **Annotations_TSVTest**,**Annotations_CSVTest**, **Text_cells**,**Empty2**

@rodichenko @mzueva 
***Verified***

**Comment by Tatyana2022** [comment]
**Bug was found:**
The selection of the previously selected heatmap disappears in Browser panel when a new file is selected in the Datasets

**To reproduce:**
1. Go to NGB
2. Go to **DATASETS** panel
3. Go to **heatmap** dataset
4. Select **tree**  file
5. Go to **BROWSER** panel
6. Select **tree** file
7. Go back to **DATASETS** panel
7. Select **large** file
8. Look at heatmap visual block in browser panel

**Expected result:**
The new selected heatmap should be added to the heatmap list and should be unchecked (**large**) 
The previously chosen checkbox should be selected(**tree** heatmap)

**Actual result:**
All heatmaps are not selected

@mzueva @rodichenko 
***Verified***

**Comment by Tatyana2022** [comment]
**Bug was found:**
An error appears in the console when double-clicking on the 'large' heatmap in BROWSER panel

**Preconditions:**
Open Dev-Console (Ctrl+F12)

**To reproduce:**
1. Go to **DATASETS** panel
2. Select **large** file in **heatmap** dataset
3. Go to **Browser** panel
4. Set checkbox in **large** value in **HEATMAP** section
5. Go to **DATASETS** panel
6. Unset checkbox for **large** file
7. Look at **HEATMAP** section in **BROWSER** panel

**Expected result:**
No errors in the console

**Actual result:**
An error displays in the console
<img src="https://user-images.githubusercontent.com/83421565/144040859-5c69de12-8bd0-4a0d-9b5d-6db48c201731.png" width="200" />

`app.bundle.js:690 Uncaught TypeError: Cannot read properties of undefined (reading 'size')
    at RowsRenderer.value (app.bundle.js:690)
    at RowsRenderer.value (app.bundle.js:690)
    at app.bundle.js:690`

**Extra Details:**
The same behavior for **gene-heatmap**
@rodichenko  @mzueva

**Comment by Tatyana2022** [comment]
**Bug was found:**
Сhromosome values are not displayed in the color scheme

**To reproduce:**
1. Go to NGB
2. Go to **DATASETS** panel
3. Go to **heatmap** dataset
4. Select **Annotations_CSVTest**  file
5. Go to **BROWSER** panel
6. Set checkbox in **Annotations_CSVTest**
7. Open **Color Scheme** window

**Expected result:**
Each value in a cell should be assigned a specific color

**Actual result:**
The color is assigned to all values, except for the values with the coordinates of the chromosomes. 
Instead of coordinates, empty fields with warning message is displayed
<img src="https://user-images.githubusercontent.com/83421565/144043133-89fce22c-9931-49b3-a997-70e9188d76d0.png" width="200" />
<img src="https://user-images.githubusercontent.com/83421565/144043497-95d0732e-8a1b-4b54-948d-7ce795164455.png" width="200" />


@mzueva @rodichenko
